transfers/reingest.py

```diff
@@ -0,0 +1,491 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+Reingest AIPs Automatically.
+
+Created for CCArch to reingest compressed AIPs using an alternative
+Archivematica ProcessingMCP.xml file.
+
+A work in progress, with some improvements that can be made to long-running
+processes like this over time.
+"""
+from __future__ import print_function
+
+import argparse
+import atexit
+import json
+import logging
+import os
+import time
+import sys
+
+
+# Allow execution as an executable and the script to be run at package level
+# by ensuring that it can see itself.
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+from transfers import errors, loggingconfig
+from transfers.amclient import AMClient
+from transfers import reingestmodel as reingestunit
+
+LOGGER = logging.getLogger('transfers')
+
+# Used early in the script to indicate failure with non-zero status, i.e.
+# something went wrong.
+ERR_PROCESSING = 1
+
+# A sleep value to realistically allow for reingest to start and approval of
+# reingest to happen.
+LATENCY = 0.8
+
+# If the process is running already we don't want to atexit to execute with
+# its default registered behavior. Override here.
+OVERRIDE_ATEXIT = False
+
+
+def setup_reingest(config):
+    """Setup the script for reingest."""
+    with open(config) as json_data_file:
+        data = json.load(json_data_file)
+        # The function atexit will ensure the PID is removed if one is created.
+        # We need to make sure that config is loaded first for this to happen.
+        atexit.register(manage_process, data, remove=True)
+        return data
+
+
+def setup_amclient(amclient):
+    """Setup the AM Client Class.
+
+    Setup the amclient class that we're going to use. We set the attributes
+    hereto avoid a Python AttributeError when using dot syntax.
+    We avoid having to use setattr(...) by doing this. Once amclient has been
+    refactored so that Issue #71 is addressed we should be able to delete
+    this block. Setattr(...) calls in this block are ordered alphabetically.
+    """
+    setattr(amclient, "aip_uuid", None)
+    setattr(amclient, "package_uuid", None)
+    setattr(amclient, "pipeline_uuid", None)
+    setattr(amclient, "processing_config", None)
+    setattr(amclient, "sip_uuid", None)
+    setattr(amclient, "transfer_directory", None)
+    setattr(amclient, "transfer_uuid", None)
```

Comment:
Can you make sure the docstring in this function is accurate? You write "We avoid having to use setattr(...)". Do you mean "We avoid having to use getattr(...)"? Small typo in "hereto". Also, why aren't you just using dot notation here, e.g., 

    amclient.aip_uuid = None
    amclient.package_uuid = None

etc.

Reply:
I can update the doc string. The reason we're not using dot notation here is what should be explained more clearly - if the attribute isn't set in the class via initialisation (or if they're not already member variables) then we get an attribute error. That's what I see when i test it that way. I thought it was clearer, but if there's some doubt it might be worth checking out the behaviour for yourself too.

Reply:
I'm sorry, I'm still a bit puzzled by this. As far as I know, `setattr(amclient, "sip_uuid", None)` and `amclient.sip_uuid = None` are directly equivalent. The only reason to use `setattr` is when the name of the attribute is only known from a variable. Is that not true? If Postel would be interjecting here, feel free to ignore :)